src/rust/engine/fs/fs_util/src/main.rs

```diff
@@ -399,59 +399,95 @@ fn paths_of_child_dir(paths: Vec<fs::PathStat>) -> Vec<fs::PathStat> {
     .collect()
 }
 
-///
-/// TODO: Stop `wait()`ing and go fully parallel.
-///
 fn materialize_directory(
   store: Arc<Store>,
-  destination: &Path,
+  destination: PathBuf,
   fingerprint: Fingerprint,
-) -> Result<(), ExitError> {
-  let directory = store.load_directory_proto(fingerprint).wait()?.ok_or_else(
-    || {
-      ExitError(
-        format!("Directory with fingerprint {} not found", fingerprint),
-        ExitCode::NotFound,
-      )
-    },
-  )?;
-  make_clean_dir(&destination).map_err(|e| {
+) -> BoxFuture<(), ExitError> {
+  let mkdir = make_clean_dir(&destination).map_err(|e| {
     format!(
       "Error making directory {:?}: {}",
       destination,
       e.description()
-    )
-  })?;
-  for file_node in directory.get_files() {
-    let fingerprint = Fingerprint::from_hex_string(&file_node.get_digest().get_hash())?;
-    match store.load_file_bytes(fingerprint).wait()? {
-      Some(bytes) => {
-        let path = destination.join(file_node.get_name());
-        File::create(&path)
-          .and_then(|mut f| f.write_all(&bytes))
-          .map_err(|e| {
-            format!("Error writing file {:?}: {}", path, e.description())
-          })?;
-      }
-      None => {
-        return Err(ExitError(
-          format!(
-            "File with fingerprint {} not found",
-            file_node.get_digest().get_hash()
-          ),
+    ).into()
+  });
+  match mkdir {
+    Ok(()) => {}
+    Err(e) => return future::result(Err(e)).to_boxed(),
+  };
+  store
+    .load_directory_proto(fingerprint)
+    .map_err(|e| e.into())
+    .and_then(move |directory_opt| {
+      directory_opt.ok_or_else(|| {
+        ExitError(
+          format!("Directory with fingerprint {} not found", fingerprint),
           ExitCode::NotFound,
-        ))
-      }
-    }
-  }
-  for directory_node in directory.get_directories() {
-    materialize_directory(
-      store.clone(),
-      &destination.join(directory_node.get_name()),
-      Fingerprint::from_hex_string(directory_node.get_digest().get_hash())?,
-    )?;
-  }
-  Ok(())
+        )
+      })
+    })
+    .and_then(move |mut directory| {
+      let file_futures = directory
+        .take_files()
+        .into_iter()
+        .map(|file_node| {
+          let path = destination.join(file_node.get_name());
+          materialize_file(store.clone(), path, file_node)
```

Comment:
Seems odd that this takes a FileNode rather than a Fingerprint.

(And similarly odd that this `take_files()`s and `into_iter()`s rather than just iterating over a read-only view, which I think it only does because you're passing a FileNode here rather than a Fingerprint)

Reply:
Good call... this looked necessary before I extracted the method, but was not. This improved the symmetry quite a bit.